File: src/core/blockchain.js
const { Transaction } = require('./transaction');
const { Consensus } = require('./consensus');
const crypto = require('crypto');

class Blockchain {
    constructor() {
        this.chain = [];
        this.pendingTransactions = [];
        this.eventLog = [];
        this.createGenesisBlock();
        this.consensus = new Consensus(this);
    }

    createGenesisBlock() {
        const genesisBlock = {
            index: 0,
            timestamp: Date.now(),
            transactions: [],
            previousHash: '0',
            hash: this.calculateHash(0, [], '0'),
        };
        this.chain.push(genesisBlock);
    }

    calculateHash(index, transactions, previousHash) {
        return crypto
            .createHash('sha256')
            .update(index + JSON.stringify(transactions) + previousHash)
            .digest('hex');
    }

    createTransaction(sender, recipient, amount, signature) {
        const transaction = new Transaction(sender, recipient, amount, signature);
        transaction.validate();
        this.pendingTransactions.push(transaction);
        this.logEvent(`Transaction created: ${JSON.stringify(transaction)}`);
        return this.getLastBlock().index + 1; // Return the next block index
    }

    getLastBlock() {
        return this.chain[this.chain.length - 1];
    }

    minePendingTransactions(minerAddress) {
        if (this.pendingTransactions.length === 0) {
            throw new Error('No transactions to mine');
        }

        const block = {
            index: this.chain.length,
            timestamp: Date.now(),
            transactions: this.pendingTransactions,
            previousHash: this.getLastBlock().hash,
            hash: this.calculateHash(this.chain.length, this.pendingTransactions, this.getLastBlock().hash),
        };

        this.chain.push(block);
        this.pendingTransactions = [];
        this.consensus.performConsensus(); // Call consensus algorithm after mining
        this.logEvent(`Block mined: ${JSON.stringify(block)}`);
        return block;
    }

    logEvent(event) {
        this.eventLog.push({ timestamp: Date.now(), event });
    }

    validateChain() {
        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            // Check if the hash is correct
            if (currentBlock.hash !== this.calculateHash(currentBlock.index, currentBlock.transactions, previousBlock.hash)) {
                return false;
            }

            // Check if the previous hash is correct
            if (currentBlock.previousHash !== previousBlock.hash) {
                return false;
            }
        }
        return true;
    }

    getChain() {
        return this.chain;
    }

    getEventLog() {
        return this.eventLog;
    }

    getPendingTransactions() {
        return this.pendingTransactions;
    }

    // Smart contract execution placeholder
    executeSmartContract(contractCode, params) {
        // This is a placeholder for executing smart contracts
        // In a real implementation, you would use a VM (like Ethereum's EVM) to execute the contract
        console.log(`Executing smart contract: ${contractCode} with params: ${JSON.stringify(params)}`);
        // Return a mock result
        return { success: true, result: 'Contract executed successfully' };
    }
}

module.exports = Blockchain;

File: decentralized_ai_integration/NeuroNet/src/blockchain/blockchain.py
import hashlib
import json
from time import time
from transaction import Transaction

class Block:
    def __init__(self, index, previous_hash, timestamp, transactions):
        self.index = index
        self.previous_hash = previous_hash
        self.timestamp = timestamp
        self.transactions = transactions
        self.hash = self.calculate_hash()

    def calculate_hash(self):
        block_string = json.dumps(self.__dict__, sort_keys=True).encode()
        return hashlib.sha256(block_string).hexdigest()

class Blockchain:
    def __init__(self):
        self.chain = []
        self.current_transactions = []
        self.create_block(previous_hash='1')  # Genesis block

    def create_block(self, previous_hash):
        block = Block(
            index=len(self.chain) + 1,
            previous_hash=previous_hash,
            timestamp=time(),
            transactions=self.current_transactions
        )
        self.current_transactions = []  # Reset the current transactions
        self.chain.append(block)
        return block

    def add_transaction(self, transaction):
        self.current_transactions.append(transaction)
        if len(self.current_transactions) >= 1:  # Example: create a block after 1 transaction
            self.create_block(previous_hash=self.chain[-1].hash)

    def to_json(self):
        return json.dumps([block.__dict__ for block in self.chain], default=str)
